keps/sig-cluster-lifecycle/kubeadm-config-ha-support.md

```diff
@@ -0,0 +1,147 @@
+
+#### Adding Additional Master-Specific ConfigMaps
+
+The proposed solution is to add additional kubeadm ConfigMaps that are specific to each master (one ConfigMap for each master).
+Each master-specific ConfigMap will be created as part of the to-be-implemented [`kubeadm join --master` process](https://github.com/kubernetes/community/pull/1707).
+Any master-specific information in the main `kubeadm-config` ConfigMap will be removed.
+
+The names of these new ConfigMaps will be `kubeadm-config-<machine_UID>` where `machine_UID` is an identifier that is guaranteed to be unique for each node in the cluster.
+There is a precedent for using such a `machine_UID`, and in fact kubeadm already has a [prerequisite](https://kubernetes.io/docs/setup/independent/install-kubeadm/#verify-the-mac-address-and-product_uuid-are-unique-for-every-node) that such machine identifiers be unique for every node.
+For the purpose of this KEP, let us assume that `machine_UID` is the full `product_uuid` of the machine that the master node is running on.
+
+Kubeadm operations such as upgrade that require master-specific information should now also grab the corresponding ConfigMap for their node.
+
+##### Key Design Considerations and Benefits
+
+There are a few key benefits to the approach of adding additional ConfigMaps over an approach which would augment the existing `kubeadm-config` with master-specific information:
+
+###### Parallel Node Creation
+
+Node creation in parallel is a valid use-case that works today.
+If we required that each master modify `kubeadm-config`, then there may be concern around concurrent modification of this ConfigMap.
+
+###### Guaranteed Consistent kubeadm-config
+
+This approach allows us to continue to guarantee that the main `kubeadm-config` is consistent with the actual cluster configuration.
+If we put master-specific information into `kubeadm-config` itself, then we would require either a yet-to-be-defined `kubeadm leave` workflow or active reconciliation of `kubeadm-config` in order to ensure accurateness.
+This may not be critical, but it is a consideration.
+
+With this proposal, if a node unexpectedly leaves a cluster, then at worst a stale ConfigMap will be left in the cluster.
+For the case where a node is explicitly deleted, we can leverage garbage collection to automatically delete the master-specific ConfigMap by listing the node as an `ownerReference` when the ConfigMap is created.
+
+### Risks and Mitigations
+
+There will be situations in which a kubeadm operation (e.g. upgrade) that requires the new master-specific ConfigMap is run and finds that the expected ConfigMap does not exist.
+For example, this will happen for users who are upgrading HA clusters that were created using the aforementioned workarounds required before kubeadm HA support is available.
+In this case, there are several options:
+
+1. Kubeadm can fall back to looking at the main `kubeadm-config`.
```

Comment:
Can't you just:
1. scan for existing non_UID configs 
2. read in and check local name = name in the config, if so delete old and drop new. 
3. if not, create new one.

Reply:
+1 
We should ensure an automatic upgrade flow for single--master nodes
For HA cluster, created with manual workaround, IMO It is acceptable to document  some manual step as well

Reply:
Yes, I believe we can automate this actually, which is great.

I was originally thinking it would be difficult / impossible due to not being able to decide whether the existing `NodeName` refers to the current node or not, given the lack of a UID in `kubeadm-config`. However, we can do this:

1. Figure out which node we are by listing all master nodes in the cluster and looking at the existing `product_uuid` field
2. Pull the `NodeName` from our node's metadata
3. Do as @timothysc describes in his comment

Or maybe there's an even simpler way to determine the current node's `NodeName` that I'm overlooking?